pkg/collector/upgrade/v0_61_0.go

```diff
@@ -0,0 +1,66 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package upgrade
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/open-telemetry/opentelemetry-operator/apis/v1alpha1"
+	"github.com/open-telemetry/opentelemetry-operator/pkg/collector/adapters"
+)
+
+func upgrade0_61_0(u VersionUpgrade, otelcol *v1alpha1.OpenTelemetryCollector) (*v1alpha1.OpenTelemetryCollector, error) {
+	if len(otelcol.Spec.Config) == 0 {
+		return otelcol, nil
+	}
+
+	otelCfg, err := adapters.ConfigFromString(otelcol.Spec.Config)
+	if err != nil {
+		return otelcol, fmt.Errorf("couldn't upgrade to v0.61.0, failed to parse configuration: %w", err)
+	}
+
+	// Search for removed Jaeger remote sampling settings. (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14163)
+	receiversConfig, ok := otelCfg["receivers"].(map[any]any)
+	if !ok {
+		// In case there is no extensions config.
+		return otelcol, nil
+	}
+
+	for key, rc := range receiversConfig {
+		k, ok := key.(string)
+		if !ok {
+			continue
+		}
+		cfg, ok := rc.(map[any]any)
+		// check if jaeger is configured
+		if !ok || !strings.HasPrefix(k, "jaeger") {
+			continue
+		}
+
+		// check if remote sampling settings exit
+		if _, ok := cfg["remote_sampling"]; !ok {
+			continue
+		}
+
+		const issueID = "https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14707"
+		return nil, fmt.Errorf(
+			"jaegerremotesampling is no longer available as receiver configuration. "+
```

Comment:
I like that the message points to the GH issue, but the root cause is not obvious. I would suggest putting either here a sentence that the ports should be changed for receiver and extension or make it super clear in the GH issue.

Reply:
Second question:

This is the first time when we actually expect the upgrade to fail. How is this error propagated to the end-user?  It would be good to have this in the operator logs but as well recoded as k8s evet. 

Third question: What happens if the upgrade fails? The instance version in the status will not be upgraded/the image will not be upgraded? If the user fixes the CR by splitting the ports will the upgrade kick in again?

Reply:
> make it super clear in the GH issue

i tried to make it clear by saying: `The jaeger remote sampling extension collides when converting old jaeger receiver settings. The problem can be solved by manually switching to another port, but this leads to potentially invalid configurations for reporters.` What would you recommend instead? Or what error message do you have in mind?

> This is the first time when we actually expect the upgrade to fail. How is this error propagated to the end-user? It would be good to have this in the operator logs but as well recoded as k8s evet.

Good point. All actions are simply [ignored](https://github.com/open-telemetry/opentelemetry-operator/blob/v0.60.0/pkg/collector/upgrade/upgrade.go#L64). Right now there is only a single [log entry](https://github.com/open-telemetry/opentelemetry-operator/blob/v0.60.0/pkg/collector/upgrade/upgrade.go#L118).

> What happens if the upgrade fails? The instance version in the status will not be upgraded/the image will not be upgraded? If the user fixes the CR by splitting the ports will the upgrade kick in again?

Your CR status will not change. If you fix this issue manually, its all fine. There is no other upgrade routine for `0.61.0`. Everytime you restart or deploy a newer operator version, all upgrade routines are triggered again.

Reply:
> Your CR status will not change. If you fix this issue manually, its all fine. There is no other upgrade routine for 0.61.0. Everytime you restart or deploy a newer operator version, all upgrade routines are triggered again.

After the fix is done, will the operator upgrade the instance by bumping the OTELcol version?

Reply:
I need to double check that. But i assume when you modified your CR, yes.

Reply:
Please report back if the upgrade was tested when Jaeger receiver was used with enabled remote sampling. We need to make sure the instance will be upgraded after the CR is fixed (e.g. extension mapped to a different port)

Reply:
@pavolloffay i have now been able to test it locally. after a configuration update, the version is **not** automatically raised. To achieve this, either the operator must be restarted or the collector configuration must be deleted and recreated.

<details>
  <summary>Details...</summary>

Operator log message, once the config is detected.
```bash
1.665433033457368e+09	ERROR	collector-upgrade	failed to upgrade managed otelcol instances	{"name": "simplest", "namespace": "test", "error": "jaegerremotesampling is no longer available as receiver configuration. Please use the extension instead with a different remote sampling port. See: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14707"}
```

procedure:
```
# status
$ kubectl get opentelemetrycollectors.opentelemetry.io -n test simplest        
NAME       MODE         VERSION   AGE
simplest   deployment   0.60.0    11m

# fix config
$ kubectl apply -n test -f v0_60_0_jaeger_cfg.yaml                      
opentelemetrycollector.opentelemetry.io/simplest configured

# status again
$ kubectl get opentelemetrycollectors.opentelemetry.io -n test simplest      
NAME       MODE         VERSION   AGE
simplest   deployment   0.60.0    12m

# delete
$ kubectl delete -n test -f v0_60_0_jaeger_cfg.yaml                
opentelemetrycollector.opentelemetry.io "simplest" deleted

# create
$ kubectl apply -n test -f v0_60_0_jaeger_cfg.yaml 
opentelemetrycollector.opentelemetry.io/simplest created

# status
$ kubectl get opentelemetrycollectors.opentelemetry.io -n test simplest 
NAME       MODE         VERSION   AGE
simplest   deployment   0.61.0    3s
```
</details>

Should we simply add that to the issue description, add a auto update flag or do you have other ideas?

Reply:
That is a good finding, we should definitely document this in the codebase that if the error is thrown during the upgrade then the version is never updated. 

There are several ways how we can proceed:
* document the breaking change and how to resolve it
* trigger the upgrade routine during the reconciliation logic

Reply:
In case of a failing update, we will now create a log entry and event. See: https://github.com/open-telemetry/opentelemetry-operator/pull/1116/commits/557a3ffd2a5f3d07f2a245adc1859c2559c15835